api/service.go

```diff
@@ -524,6 +528,158 @@ func (s *service) RemoveAppToken(ctx context.Context, req *pb.RemoveAppTokenRequ
+	log.Debugf("received add folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	pth, err := s.ipfsClient.Unixfs().Add(
+		ctx,
+		ipfsfiles.NewMapDirectory(map[string]ipfsfiles.Node{}),
+		options.Unixfs.Pin(true))
+	if err != nil {
+		return nil, err
+	}
+
+	folder, err := s.collections.Folders.Create(ctx, pth, req.Name, req.Public, proj.ID)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.AddFolderReply{
+		ID:   folder.ID,
+		Path: folder.Path,
+	}, nil
+}
+
+// GetFolder handles a get folder request.
+func (s *service) GetFolder(ctx context.Context, req *pb.GetFolderRequest) (*pb.GetFolderReply, error) {
+	log.Debugf("received get folder request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	folder, err := s.getFolderWithScope(ctx, req.Name, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+	if err != nil {
+		return nil, err
+	}
+	defer node.Close()
+
+	return folderToPbFolder(folder, node)
+}
+
+func folderToPbFolder(folder *c.Folder, node ipfsfiles.Node) (*pb.GetFolderReply, error) {
+	entries, err := dirToPbSlice(node)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.GetFolderReply{
+		ID:        folder.ID,
+		Path:      folder.Path,
+		Name:      folder.Name,
+		Public:    folder.Public,
+		ProjectID: folder.ProjectID,
+		Created:   folder.Created,
+		Entries:   entries,
+	}, nil
+}
+
+func dirToPbSlice(node ipfsfiles.Node) (entries []*pb.GetFolderReply_Entry, err error) {
+	err = ipfsfiles.Walk(node, func(path string, n ipfsfiles.Node) error {
+		defer n.Close()
+		if path == "" { // This is the root
+			return nil
+		}
+		size, err := n.Size()
+		if err != nil {
+			return err
+		}
+		entry := &pb.GetFolderReply_Entry{
+			Path: path,
+			Size: size,
+		}
+		switch n.(type) {
+		case ipfsfiles.Directory:
+			entry.IsDir = true
+		}
+		entries = append(entries, entry)
+		return nil
+	})
+	return
+}
+
+// ListFolders handles a list folders request.
+func (s *service) ListFolders(ctx context.Context, req *pb.ListFoldersRequest) (*pb.ListFoldersReply, error) {
+	log.Debugf("received list folders request")
+
+	scope, ok := ctx.Value(reqKey("scope")).(string)
+	if !ok {
+		log.Fatal("scope required")
+	}
+	proj, err := s.getProjectForScope(ctx, req.ProjectID, scope)
+	if err != nil {
+		return nil, err
+	}
+
+	folders, err := s.collections.Folders.List(ctx, proj.ID)
+	if err != nil {
+		return nil, err
+	}
+	list := make([]*pb.GetFolderReply, len(folders))
+	for i, folder := range folders {
+		node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+		if err != nil {
+			return nil, err
+		}
+		list[i], err = folderToPbFolder(folder, node)
+		if err != nil {
+			return nil, err
+		}
+		node.Close()
```

Comment:
Should be `defer`ed and moved above to be called on early returns?

Reply:
Ah yeah, good catch. I'd opt for just calling it explicitly in the return case to avoid the `defer` in loop, sound okay?

Reply:
Yep.